nwb_conversion_tools/utils/basenwbephyswriter.py

```diff
@@ -0,0 +1,161 @@
+import pynwb
+import numpy as np
+import uuid
+from datetime import datetime
+
+
+class BaseNwbEphysWriter:
+    def __init__(self, object_to_write, nwb_file_path=None, nwbfile=None, metadata=None, **kwargs):
+        self.object_to_write = object_to_write
+        self.nwb_file_path = nwb_file_path
+        self.metadata = metadata
+        self.nwb_file_path = nwb_file_path
+        self.nwbfile = nwbfile
+        self._kwargs = kwargs
+
+    def instantiate_nwbfile(self):
+        # Default arguments will be over-written if contained in metadata
+        nwbfile_kwargs = dict(
+            session_description="Auto-generated by NwbRecordingExtractor without description.",
+            identifier=str(uuid.uuid4()),
+            session_start_time=datetime(1970, 1, 1),
+        )
+        if self.metadata is not None and "NWBFile" in self.metadata:
+            nwbfile_kwargs.update(self.metadata["NWBFile"])
+        self.nwbfile = pynwb.NWBFile(**nwbfile_kwargs)
+
+    @staticmethod
+    def get_kwargs_description(self):
+        raise NotImplementedError
+
+    def write_to_nwb(self):
+        raise NotImplementedError
+
+    def write_recording(self):
+        raise NotImplementedError
+
+    def write_sorting(self):
+        raise NotImplementedError
+
+    def write_epochs(self):
+        raise NotImplementedError
+
+    def write_waveforms(self):
+        raise NotImplementedError
+
+    def get_nwb_metadata(self):
+        raise NotImplementedError
+
+    def add_devices(self):
+        """
+        Auxiliary static method for nwbextractor.
+
+        Adds device information to nwbfile object.
+        Will always ensure nwbfile has at least one device, but multiple
+        devices within the metadata list will also be created.
+
+        Missing keys in an element of metadata['Ecephys']['Device'] will be auto-populated with defaults.
+        """
+        if self.nwbfile is not None:
+            assert isinstance(self.nwbfile, pynwb.NWBFile), "'nwbfile' should be of type pynwb.NWBFile"
+
+        # Default Device metadata
+        defaults = dict(name="Device", description="Ecephys probe. Automatically generated.")
+
+        if self.metadata is None:
+            self.metadata = dict()
+
+        if "Ecephys" not in self.metadata:
+            self.metadata["Ecephys"] = dict()
+
+        if "Device" not in self.metadata["Ecephys"]:
+            self.metadata["Ecephys"]["Device"] = [defaults]
+
+        for dev in self.metadata["Ecephys"]["Device"]:
+            if dev.get("name", defaults["name"]) not in self.nwbfile.devices:
+                self.nwbfile.create_device(**dict(defaults, **dev))
+
+    def add_electrodes(self):
```

Comment:
what is the difference between `write_*` and `add_*` functions?

Reply:
The `write_*` functions are in charge of organizing the writing of big portions of the NWB file (e.g. `write_recording`). The add functions are helper functions that add small bits of info to existing NWB files instances.